### PR TITLE
#3231 RTE editor in grid collapsed on tab shown

### PR DESCRIPTION
**Pull request by agrath**
Fixes https://github.com/umbraco/Umbraco-CMS/issues/3231

### CODE REVIEW

**Comment by poornimanayar** [comment]
Hi @agrath ,

Many thanks for your PR. I shall give it a spin today and let you know how this goes :-)

Regards,
Poornima, [Community Pull Request team](https://umbraco.com/blog/introducing-the-new-cpr-team/)

**Comment by poornimanayar** [comment]
Hi @agrath ,

I have managed to test this. This is what I did

- Set up a doc type with a grid on 2nd tab
- Created a content node and added RTE to the grid with some content
- Moved back to first tab
- Went back to the second tab after 20s or more

This is what I see following the steps above, can you confirm that I am on the right track?
**Before**
image

**After**
image

Can I request that you add the ticket details and steps to reproduce/supported screenshots etc to the PR description going forward? Just helps us with testing :)

Regards,
Poornima, [Community Pull Request team](https://umbraco.com/blog/introducing-the-new-cpr-team/)

**Comment by agrath** [comment]
Hi @poornimanayar 

By before/after do you mean before & after patch is applied? If so, then yes these are the correct steps to reproduce the fault & test the solution.
The RTE content not displaying in the grid until you click on it (after visiting the tab after 20 second timeout elapses) is the fault that this PR fixes.
In my testing, I had the doc types set up the same, but then just stayed on the first tab for 20 seconds before switching to the second tab.
What's happening in the background is the RTE controls in the grid are doing an autoresize (which uses a timeout loop, up to 20 times, 1000ms apart) but result in a height of 16px because the RTE is invisible (on a non visible tab)
When you switch to the tab, then you can't see the content - so this fix just makes the RTE reattempt the auto-resize after the tab is shown.
Hope this makes sense

**Comment by nul800sebastiaan** [comment]
Good point @dawoe! @agrath could you adjust it a little bit to use `eventService` instead? Other than that, this looks really good!

**Comment by agrath** [comment]
@nul800sebastiaan @dawoe I have updated my branch to use the eventsService rather than $rootScope directly

**Comment by nul800sebastiaan** [comment]
Thanks @agrath - works great ! 👍

**Comment by nul800sebastiaan** [comment]
Hi there @agrath,

First of all: **A big #H5YR** for contributing to Umbraco during Hacktoberfest! We are very thankful for the huge amount of PRs submitted, and all the amazing work you've been doing :pray:

If you haven't heard already we have had a record breaking amount of PR's in both the CMS and the Documentation repository. You can read all about the numbers in the [Umbraco Hacktoberfest Aftermath blog post](https://umbraco.com/blog/the-2018-hacktoberfest-aftermath/).

In the spirit of Hacktoberfest we've prepared some **exclusive Umbraco swag** for all our contributors - _including you_! 

[Receive your swag!](https://goo.gl/forms/smjQwEMDM5cUnz6x1) :point_left: Please follow this link to fill out and submit the form, **before November 21st 2018**.

Following this date we'll be sending out all the swag, which also means that it might not reach your doorstep before December, so please bear with us and be patient :pray:

The only thing left to say is **thank you so much for participating in Hacktoberfest**! We really appreciate the help!

Kind regards,
The Umbraco PR Team & The Documentation Curators

File: src/Umbraco.Web.UI.Client/src/common/directives/validation/valtab.directive.js [hunk 1]
```diff
@@ -31,8 +31,17 @@ function valTab() {
                     scope.tabHasError = false;
                 }
             });
+            var tabShownFunc = function (e) {
+                var tabContent = element.closest(".umb-panel").find("#" + tabId);
+                eventsService.emit('valTab.tabShown', { originalEvent: e, tab: scope.tab, content: tabContent });
```
Comment (Shazwazza) [inline src/Umbraco.Web.UI.Client/src/common/directives/validation/valtab.directive.js]: This logic should not exist here. This directive is purely to do with helping out with validation. There is already an event for switching tabs which is controlled by the tabs themselves in `umbraco.directives.directive:umbTabs` and the event is `app.tabChange`. 

I'll remove this and change the logic for this PR.

Comment (Shazwazza) [reply src/Umbraco.Web.UI.Client/src/common/directives/validation/valtab.directive.js]: Fixed in rev 9ae4d1cdf50625cd7c55f6fdb917b6dc45f253f4

**Comment by perplexErik** [comment]
Hi,

Some of our customers are reporting the loss of content from the grid so i immediately tried the changes to the .js files as described in  #3231 ,but the issue remains.
(Umbraco version 7.10.4 assembly: 1.0.6684.15725)

What i tried:

- applied changes in file: \Umbraco\Js\umbraco.directives.js

- Deleted the /App_Data\TEMP\ClientDependency folder

- Incremented the version number in \Config\ClientDependency.config

- Touched the the web.config file

- Cleared Google Chrome cache

After doing the above, i can still reproduce the issue.
I create 3 grid columns, Add 3 RTE editors which some text and 1 image element.
Select "Reorder" and start shuffling the RTE and image blocks between columns.
After that "select i am done reordering" and 1 of the RTE's content is no longer visible.

Did i overlook something?

Kind regards,
Erik

**Comment by agrath** [comment]
@perplexErik it sounds like a very similar bug, 
in `src/common/directives/components/grid/grid.rte.directive.js` we would need to subscribe to the resorted event (unsure which this is) and raise the same reinitialize mce command - that should fix it.
Can you report this as a seperate issue? it would also be great if you could supply a standalone umbraco site with the content all set up ready for testing.

**Pull request merged**